specs/fault-proof-vm.md

```diff
@@ -0,0 +1,177 @@
+
+## Overview
+
+This is a description of the Fault Proof Virtual Machine (FPVM). The FPVM emulates a minimal Linux-based system
+running on big-endian 32-bit MIPS32 architecture. Alot of its behaviors are copied from Linux/MIPS with a few
+tweaks made for fault proofs.
+
+Operationally, the Fault Proof VM is a state transition function. This state transition is referred to as a *Step*,
+that executes a single instruction. We say the VM is a function $f$, given an input state $S_{pre}$, steps on a
+single instruction encoded in the state to produce a new state $S_{post}$.
+$$f(S_{pre}) \rightarrow S_{post}$$
+
+Thus, the trace of a program executed by the FPVM is an ordered set of VM states.
+
+## State
+
+The virtual machine state highlights the effects of running a Fault Proof Program on the VM.
+It consists of the following fields:
+
+1. `memRoot` - A `bytes32` value representing the merkle root of VM memory.
+2. `preimageKey` - `bytes32` value of the last requested pre-image key.
+3. `preimageOffset` - The 32-bit value of the last requested pre-image offset.
+4. `pc` - 32-bit program counter.
+5. `nextPC` - 32-bit next program counter. Note that this value may not always be $pc+4$
+ when executing a branch/jump delay slot.
+6. `lo` - 32-bit MIPS LO special register.
+7. `hi` - 32-bit MIPS HI special register.
+8. `heap` - 32-bit address of the base of the VM heap memory.
+9. `exitCode` - 8-bit exit code.
+10. `exited` - 1-bit indicator that the VM has exited.
+11. `registers` - General-purpose MIPS32 registers. Each register is a 32-bit value.
+
+The state is represented by packing the above fields, in order, into a 226-byte buffer.
+
+## Memory
+
+Memory is represented as a binary merkle tree.
+The tree has a fixed-depth of 27 levels, with leaf values of 32 bytes each.
+This spans the full 32-bit address space, where each leaf contains the memory at that part of the tree.
+The state `memRoot` represents the merkle root of the tree, reflecting the effects of memory writes.
+As a result of this memory representation, all memory operations are 4-byte aligned.
+Memory access doesn't require any privileges. An instruction step can access any memory
+location as the entire address space is unprotected.
+
+### Heap
+
+FPVM state contains a `heap` that tracks the current address of the free store used for memory allocation.
+Heap pages are bump allocated at the page boundary, per `mmap` syscall. The page size is 4096.
+
+The FPVM has a fixed program break at `0x40000000`. However, the FPVM is permitted to extend the
+heap beyond this limit via mmap syscalls.
+For simplicity, there are no memory protections against "heap overruns" against other memory segments.
+Such VM steps are still considered valid state transitions.
+
+Specification of memory mappings is outside the scope of this document as it is irrelevant to
+the VM state. FPVM implementers may refer to the Linux/MIPS kernel for inspiration.
+
+## Delay Slots
```

Comment:
Good point. The VM behavior here should be deterministic though rather than undefined. I'll update the spec to indicate that an Exception should be triggered, preventing a VM step to occur.
The Go and Solidity VMs may need to be updated to adhere to this.

Reply:
I think it can be implemented by checking if already `NextPC != PC + 4` during a jump/branch/etc. and then raising an exception. The Go compiler should never output such sequence though.